Sources/UtiliKit/ExternalMapping/MapAppURLBuilder/Implementations/WazeURLBuilder.swift

```diff
@@ -0,0 +1,146 @@
+        if let favorite = percentEncoded(favorite.rawValue) {
+            urlComponents?.addQueryItem(.init(parameter: .favorite, value: favorite))
+        }
+        return urlComponents?.url ?? baseURL
+    }
+
+    /// Generate a URL to navigation directions in the Waze app with the specified parameters
+    /// - Parameters:
+    ///   - query: Search term or address to look up on the map.
+    ///   - coordinate: Location to display or location to search near if a `query` is specified.
+    ///   - zoom: The zoom level.
+    /// - Returns: URL to the Waze app
+    static func directions(query: String? = nil,
+                           coordinate: Coordinate? = nil,
+                           zoom: Float? = nil) -> URL {
+        var urlComponents = URLComponents(string: location(query: query, coordinate: coordinate, zoom: zoom).absoluteString)
+        urlComponents?.addQueryItem(.init(parameter: .navigate, value: "yes"))
+        return urlComponents?.url ?? baseURL
+    }
+
+    /// Generate a URL to navigation directions in the Waze app with the specified parameters
+    /// - Parameters:
+    ///   - favorite: One of the user's saved locations.
+    ///   - zoom: The zoom level.
+    /// - Returns: URL to the Waze app
+    static func directions(favorite: Favorite,
+                           zoom: Float? = nil) -> URL {
+        var urlComponents = URLComponents(string: location(favorite: favorite, zoom: zoom).absoluteString)
+        urlComponents?.addQueryItem(.init(parameter: .navigate, value: "yes"))
+        return urlComponents?.url ?? baseURL
+    }
+}
+
+// MARK: MapAppURLBuilder
+
+extension WazeURLBuilder: MapAppURLBuilder {
+
+    static var zoomRange: ClosedRange<Float> { Float(6)...8192 }
+    static var launchAppURL: URL { baseURL }
+
+    static func displayLocation(at coordinate: Coordinate, zoomPercent: Float?, style: MapStyle?) -> URL {
+        location(coordinate: coordinate, zoom: zoomValue(fromPercent: zoomPercent))
+    }
+
+    static func search(for query: String, near coordinate: Coordinate?, style: MapStyle?) -> URL {
+        location(query: query, coordinate: coordinate)
+    }
+
+    static func navigate(to toAddress: String, from fromAddress: String?, via navigationMode: NavigationMode?, style: MapStyle?) -> URL {
+        directions(query: toAddress)
+    }
+}
+
+// MARK: - WazeURLBuilder.Coordinates
+
+extension WazeURLBuilder {
+    enum Favorite: String {
+        case home
+        case work
+    }
+}
+
+// MARK: - Private
+
+private extension WazeURLBuilder {
+    static var baseURL: URL {
+        guard let url = URL(string: "waze://") else {
+            fatalError("Could not build the base URL to Waze")
+        }
+        return url
+    }
+
+    static var baseURLComponents: URLComponents? {
+        URLComponents(string: baseURL.absoluteString)
+    }
+
+    static func percentEncoded(_ string: String?) -> String? {
```

Comment:
It looks like all 3 URL builders use the same logic for percent encoding. It might be good to add an internal extension on `String` to be a bit more DRY.

```swift
extension String {
    func percentEncoded() -> String? {
        self.replacingOccurrences(of: " ", with: "+").addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed)
    }
}
```